Set recursive ACP VC access to match grant recursiveness

NSeydoux opened the pull request:
This enables end-to-end tests for recursive access grants against ESS dev-next, which is the environment where they are supported. This involves a couple of things:
- Upgrading to the latest `@inrupt/solid-client`, which adds the possibility to set a recursive VC matcher updating a resource's ACR
- Updating the end-to-end tests to only run recursive access grants tests against environment supporting them
- The `@inrupt/solid-client` upgrade comes with a couple of unplanned changes, such as `cross-fetch` mock updates, and adding `@types/rdfjs__dataset` as a dependency to fix docs generation.

NSeydoux committed:
Temporary: link to local solid-client version

NSeydoux committed:
Normalize boolean literals

A JSON-LD context may map a literal to a boolean, so we normalize "true" into true for known predicates.

NSeydoux committed:
Enable end-to-end tests for recursive grants

NSeydoux committed:
Add feature locks on recursive grants tests

Environment variables are added so that recursive access grant tests only run in environments supporting them.

NSeydoux committed:
Install solid-client@1.25.0

It introduces recursive ACP update

NSeydoux committed:
Add the default inherit value in unit tests

NSeydoux committed:
Fix cross-fetch and Response mocking

NSeydoux committed:
Fix normalization unit test

NSeydoux committed:
Replace setImmediate with setTimeout

NSeydoux committed:
Temporary fix: add rdfjs types as dependency

NSeydoux committed:
fix dependency versionOC

ThisIsMissEm commented on e2e/node/e2e.test.ts:
```diff
@@ -1014,52 +1016,65 @@ describe(`End-to-end access grant tests for environment [${environment}}]`, () =
+    /* eslint-disable jest/no-standalone-expect */
+
+    // Only enable this test in environments supporting recursive access grants
+    testIf(environmentFeatures?.RECURSIVE_ACCESS_GRANTS === "true")(
```
What if we add it to our testing framework? `const featureTest = createFeatureTests("RECURSIVE_ACCESS_GRANTS", jest)`

NSeydoux replied:
That's a very good idea

ThisIsMissEm commented on src/gConsent/discover/redirectToAccessManagementUi.browser.test.ts:
```diff
@@ -114,8 +115,9 @@ describe("redirectToAccessManagementUi", () => {
         }
       );
       // Yield the event loop to make sure the blocking promises completes.
+      // FIXME: Why is setImmediate undefined in this context ?
```
But really this code is a bit weird as it's an "await for 1 tick of the event loop then grab `window.location`"

what would be better would be to spy on `window.location` so you can observe it's changes

ThisIsMissEm commented on src/gConsent/manage/approveAccessRequest.ts:
```diff
@@ -108,7 +116,9 @@ async function addVcMatcher(
       // eslint-disable-next-line camelcase
-      const updatedResource = acp_ess_2.setVcAccess(resourceInfo, accessMode);
+      const updatedResource = acp_ess_2.setVcAccess(resourceInfo, accessMode, {
+        inherit: options?.inherit ?? true,
```
We're defaulting this to true now?

NSeydoux replied:
Indeed: the Access Grants are defaulting to recursive, so the ACR structure should do the same otherwise recursive Access Grants aren't applicable.

ThisIsMissEm replied:
Is that how product is rolling this out? Shouldn't they default to non-recursive in this major release (as was previously the case), and then on the next semver major we make them recursive by default?

NSeydoux replied:
Everyone we talked to tripped over the current behavior, expecting granting access to a container to also apply on contained resources. So technically yes, but here it's more aligning with expectations.

ThisIsMissEm commented on src/gConsent/manage/getAccessGrant.test.ts:
```diff
@@ -20,17 +20,30 @@
+  return {
+    // Do no mock the globals such as Response.
+    ...crossFetch,
+    Response: crossFetch.Response,
```
```suggestion
```

this is handled by the line above, or should be.

NSeydoux replied:
Oops yeah looks like a leftover of a test 03265e0161d371884234c0951c685076bb738fc2

ThisIsMissEm commented on src/gConsent/request/issueAccessRequest.test.ts:
```diff
@@ -460,6 +471,7 @@ describe("issueAccessRequest", () => {
             normalizedAccessRequest.credentialSubject.hasConsent
               .forPersonalData[0],
           mode: normalizedAccessRequest.credentialSubject.hasConsent.mode[0],
+          inherit: "true",
```
Do we need a test case for inherit === false?

NSeydoux replied:
Never too sure: d6ad1c5fc7857e295c3292f8423a78af21fc8230

ThisIsMissEm commented on e2e/node/e2e.test.ts:
```diff
@@ -944,9 +948,7 @@ describe(`End-to-end access grant tests for environment [${environment}}]`, () =
-  // This is skipped until ESS supports recursive Access Grants
-  // eslint-disable-next-line jest/no-disabled-tests
-  describe.skip("access grants can be recursive or only apply to the target resource", () => {
+  describe("access grants can be recursive or only apply to the target resource", () => {
```
Note on the two disabled tests above, those can probably use inherit / recursive access grants to function as expected

NSeydoux replied:
I rephrased the describe, because I think it makes sense as a single block, and that's a good point, I have re-enabled them: 34f00a0d742b846075b69b360b5b6d25dd14a3ea

ThisIsMissEm reviewed (approved):
A few small comments, but LGTMO.

NSeydoux commented on package.json:
```diff
@@ -127,8 +127,9 @@
-    "@inrupt/solid-client": "^1.23.3",
+    "@inrupt/solid-client": "^1.25.0",
     "@inrupt/solid-client-vc": "^0.5.0",
+    "@types/rdfjs__dataset": "^1.0.5",
```
Absolutely

NSeydoux committed:
Remove unneeded Response reexport

NSeydoux committed:
Remove out-of-context comment

NSeydoux committed:
Add test for "false" inherit flag

NSeydoux committed:
Re-enabled skipped tests

The tests about overwriting a resource are now enabled thanks to recursive grants, because the ACR with the VC matcher is no longer deleted.

NSeydoux committed:
Fix feature detection

NSeydoux committed:
Fix end-to-end CI
